### DISCUSSION TITLE
Unable to login to Windcentrale

### DISCUSSION THREAD

**Discussion by jonesPD** [jobvk/Home-Assistant-Windcentrale#69] [Support] [answered]
Hey Job,
Sorry for the very long response time - I was offline for a bit demolishing my kitchen and preparing for a new one :-) - would still like to create a generic version for e.g. connection to Domoticz. I have tried the template you shared (thank you!)

```
import json
import requests
import boto3
from pycognito.aws_srp import AWSSRP

boto3_client = boto3.client('cognito-idp', region_name='eu-west-1')
aws = AWSSRP(username='', password='', pool_id='eu-west-1_U7eYBPrBd', client_id='715j3r0trk7o8dqg3md57il7q0', client=boto3_client)
tokens = aws.authenticate_user()
token_type = tokens['AuthenticationResult']['TokenType']
id_token = tokens['AuthenticationResult']['IdToken']
authorization_header = {'Authorization':token_type + " " + id_token}

Request_result = requests.get("https://mijn.windcentrale.nl/api/v0/livedata?projects=WND-WJ", headers=authorization_header, verify=True)
print(Request_result.json())
```

Unfortunately the result is the same: authentication error when executing aws.authenticate_user()

botocore.errorfactory.NotAuthorizedException: An error occurred (NotAuthorizedException) when calling the RespondToAuthChallenge operation: Incorrect username or password.

Since I know my credentials are ok (I've tested them once again manually logging in), it must be one of the other parameters. Any ideas/suggestions?

Thanks a lot.

**Comment by jobvk** [comment]
I have some ideas or suggestions you could try:
1. The username parameter you should fill in your email address and not a username.
2. Check in you browser inspect network or [Postman](https://www.postman.com) which APIs and parameters are being used. 
3. Where are you testing this software on? And try reinstalling Boto3 & pyCognito

Documentation Boto3: https://boto3.amazonaws.com/v1/documentation/api/latest/index.html
Documentation pyCognito: https://github.com/pvizeli/pycognito#using-awssrp

**Reply by jobvk** [reply]
Have you had more success this week?

**Reply by jonesPD** [reply]
Hi Job
Super dat je het nog eens vraagt. Ik heb inmiddels een andere cliënt id gebruikt op basis van de Google developer tool maar dat was geen oplossing. Hoe heb je de pool id destijds bepaald? Op dezelfde manier?
Groeten,
Peter

**Comment by jonesPD** [comment]
Hi Job,
Following the steps you suggested above:
1) gotcha, good to confirm, but this is what i've been doing all along
2) using Google Chrome and its developer tool, I watched the network traffic, I see there is a bunch of API calls to AWS servers (https://cognito-idp.eu-west-1.amazonaws.com/), which confirms the used authentication protocol and region. The second call contains my email address as a USERNAME, a similar but slightly different ClientId to what you proposed, a long SRP_A value (what is this?), and a ClientMetaData portal value 'ENERGY_SSP'. The response includes a Authentication Challenge with a bunge of parameters, incl. once more my email. 
The fourth call responds to the challenge, and results in a response with a bunge of tokens. When using the IdToken included in the response, I can query the Winddelen API and get a good response: https://mijn.winddelen.nl/api/v0/livedata?projects=WND-VH
{'WND-VH': {'year_production': '4064948', 'rpm': '15.5', 'total_runtime': '0', 'year_runtime': '0', 'diameter': '67', 'pulsating': '0', 'power_percentage': '64', 'power_per_share': '60', 'id': '211', 'timestamp': '1670962310', 'wind_power': '6', 'power': '581', 'wind_direction': 'ZO'}}
So this works, but unfortunately, this only works using authentication in Chrome, and re-using the IdToken in Python.

3) So this brings me to here. For testing purposes I'm running Python 3.8 on a Windows laptop and using the command line to execute the script. I've re-installed Boto3 (v1.26.28) and pyCognito (2022.11.2), but this makes no difference. 

This makes me wonder what is causing the authentication error. My user credentials themselves are correct, but have I provided them in the right style (e.g. binary, base64 encoded, string or something else)? Is the pool_id correct for me and how to find out? What is the ClientMetaData 'ENERGY_SSP'? 

Other suggestions are obviously welcome.

**Reply by jobvk** [reply]
Are you making a Domoticz component for : https://mijn.winddelen.nl or https://mijn.windcentrale.nl?

**Reply by jonesPD** [reply]
Ah ... Oops. That explains why my credentials don't work I assume. Good catch. So...up to me to figure out how to get it working for winddelen.nl I assume. 

Thanks though for your time and help so far.

**Comment by jobvk** [comment, marked as answer]
Misunderstanding between https://mijn.winddelen.nl and https://mijn.windcentrale.nl
